pkg/cmd/image/save.go

```diff
@@ -0,0 +1,90 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package image
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/images/archive"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/idutil/imagewalker"
+	"github.com/containerd/nerdctl/pkg/platformutil"
+	"github.com/containerd/nerdctl/pkg/strutil"
+	"github.com/mattn/go-isatty"
+)
+
+// Save will save one or more images to a tar archive (streamed to STDOUT by default).
+func Save(ctx context.Context, client *containerd.Client, images []string, options types.ImageSaveOptions) error {
+	if options.Output != "" {
+		f, err := os.OpenFile(options.Output, os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		options.Stdout = f
+	} else {
+		if isatty.IsTerminal(os.Stdout.Fd()) {
+			return fmt.Errorf("cowardly refusing to save to a terminal. Use the -o flag or redirect")
+		}
+	}
```

Comment:
(below is more for discussion and see other reviewers' thought).

I think we should remove `Output string` from `ImageSaveOptions`. Functionally, it represents the same thing as `Stdout io.Writer` and having both might cause inconsistency and confusing, for example:

- Even `Output != ""`, we init `Stdout = cmd.OutOrStdout()` when creating `ImageSaveOptions`, and then change it after the file writer created (i.e., here).
- Should a caller specify `Output`, or `Stdout io.Writer`?

If we remove `Output` and keep `Stdout` only:

1. `pkg/cmd` side logic doesn't need to know where the writer comes from. It just writes content to it.
2. On `cmd` side, we init `opt.Stdout = cmd.OutOrStdout()` if `output == ""` or create a file writer and use it as `opt.Stdout` (i.e., this pard of code).

Reply:
I think your thought is a pretty good idea.

Reply:
/cc @AkihiroSuda @Zheaoli

Reply:
I think it's better to make another PR to simplify the logic. I prefer keep the same code here, (one PR for one thing)